From: Shiji Yang <yangshiji66@outlook.com>
Date: Thu, 18 Jan 2024 10:10:39 +0000
Subject: [PATCH] mtd: spi-nor: introduce small buffer read hack

The ath79 SoC GPIO voltage is 2.5V. In order to be compatible
with 3.3V NOR Flash chips, most devices use resistors for voltage
matching. However, some devices (such as TP-Link WR2543N) use a
coupling capacitor to connect Flash and GPIO. This has led to a
strange behavior where the SPI bus is unable to continuously read
correct data from Flash chips. This patch workaround this issue by
limiting the data size of each SPI transfer cycle and inserting
some delay time between each read. User can add the newly designed
property 'small-buffer-read' to the dts flash node to enable this
feature.

Signed-off-by: Shiji Yang <yangshiji66@outlook.com>
---
 drivers/mtd/spi-nor/core.c | 15 +++++++++++++--
 drivers/mtd/spi-nor/core.h |  1 +
 2 files changed, 14 insertions(+), 2 deletions(-)

--- a/drivers/mtd/spi-nor/core.c
+++ b/drivers/mtd/spi-nor/core.c
@@ -1685,6 +1685,7 @@ static int spi_nor_read(struct mtd_info
 {
 	struct spi_nor *nor = mtd_to_spi_nor(mtd);
 	ssize_t ret;
+	size_t read_len;
 
 	dev_dbg(nor->dev, "from 0x%08x, len %zd\n", (u32)from, len);
 
@@ -1695,9 +1696,16 @@ static int spi_nor_read(struct mtd_info
 	while (len) {
 		loff_t addr = from;
 
+		if (unlikely((nor->flags & SNOR_F_SMALL_BUF_READ))) {
+			read_len = len > 32 ? 32 : len;
+			usleep_range(10, 20);
+		} else {
+			read_len = len;
+		}
+
 		addr = spi_nor_convert_addr(nor, addr);
 
-		ret = spi_nor_read_data(nor, addr, len, buf);
+		ret = spi_nor_read_data(nor, addr, read_len, buf);
 		if (ret == 0) {
 			/* We shouldn't see 0-length reads */
 			ret = -EIO;
@@ -2429,6 +2437,9 @@ static void spi_nor_init_flags(struct sp
 	if (of_property_read_bool(np, "broken-flash-reset"))
 		nor->flags |= SNOR_F_BROKEN_RESET;
 
+	if (of_property_read_bool(np, "small-buffer-read"))
+		nor->flags |= SNOR_F_SMALL_BUF_READ;
+
 	if (flags & SPI_NOR_SWP_IS_VOLATILE)
 		nor->flags |= SNOR_F_SWP_IS_VOLATILE;
 
--- a/drivers/mtd/spi-nor/core.h
+++ b/drivers/mtd/spi-nor/core.h
@@ -132,6 +132,7 @@ enum spi_nor_option_flags {
 	SNOR_F_SWP_IS_VOLATILE	= BIT(13),
 	SNOR_F_RWW		= BIT(14),
 	SNOR_F_ECC		= BIT(15),
+	SNOR_F_SMALL_BUF_READ	= BIT(16),
 };
 
 struct spi_nor_read_command {
